draft-ietf-tls-external-psk-importer.md

```diff
@@ -138,14 +178,46 @@ EPSKs may be imported for early data use if they are bound to protocol settings
 
+##Client ID {#client-id-description}
+This is to prevent Selfie-style reflections.
+This is only necessary in scenarios where more than two actors use the same key, including the case where a single agent will complete PSK handshakes as both the client and the server using the same key.
+The Selfie attack {{!Selfie}} abuses an underlying assumption of TLS, that only one client and one server know a PSK.
+A pair of agents that both act as both client *and* server, and use the same PSK in both roles violate this assumption.
+In this case, there are two clients and two servers that know the PSK, since each endpoint acts as both client and server.
+The agents are thus both vulnerable to reflection attacks where the attacker forces the agent to act as both client and server in a single connection.
+This is because neither agent can distinguish between itself and its peer.
+Recall both servers have the exact same configuration.
+Therefore even though the session specific parameters change for each session, the client cannot distinguish between a response reflected back on it by the attacker and an honest response from its peer.
+For each parameter, both servers either return the value for every session, or they return a different value for every session.
+By hashing a distinguishing `client_id` value into the PSK binder, a server that received a reflected `ClientHello` would be unable to verify the binder, and would reject the connection.
+
+For example, distinguishing strings for agents in peer-to-peer IoT device deployments could be a MAC address.
+Similarly, distinguishing strings for agents in mesh VM deployments could be namespaces.
+The decision to use this field and what it should contain MUST be agreed OOB.
+Note that, critically, this value is never sent on the wire, as this would identify the client even to passive adverseries.
+
+##Prior Contexts {#prior-contexts-description}
+In the standard case this list will be empty because the TLS connection will not be wrapped in a prior security context.
+However, if the OOB PSK was established through a protocol, or series of protocols, that provide a continuing security context, then including the channel binding for these contexts, as well as any keys established, binds the security contexts together.
+This makes it easier to reason formally about the exact properties are provided by the combined series of protocols.
+
+##Example
+As an example, consider chaining together two TLS sessions using OOB PSK importers rather than resumption.
+One could then include the prior context `<tls_channel_binding, master_secret>` in the second connection, where `tls_channel_binding` is computed by calling the TLS exporter interface of the first connection with a label that uniquely defines this particular setup.
+
+Including this channel binding in the prior contexts binds the security context of the first channel to the security context of the second channel.
```

Comment:
The [TLS 1.2 exporter][1] output is derived from the master secret.
The [TLS 1.3 exporter][2] output is derived from the [early_]exporter_master_secret which is [derived][3] from the Early Secret or Master Secret.

Unless you do not trust the hash functions, I don't see why adding the master secret would help. Could you clarify this?

 [1]: https://tools.ietf.org/html/rfc5705#section-4
 [2]: https://tools.ietf.org/html/rfc8446#section-7.5
 [3]: https://tools.ietf.org/html/rfc8446#page-93